File: src/_03_BehavioralPattern/_03_08_State/PrivateState.js
import State from "./State.js";

/**
 * @typedef { import("./OnlineCourse").default } OnlineCourse
 * @typedef { import("./Student").default } Student
 */

export default class PrivateState extends State {
  /** @type { OnlineCourse } */
  _onlineCourse;

  /** @param { OnlineCourse } onlineCourse */
  constructor(onlineCourse) {
    super();
    this._onlineCourse = onlineCourse;
  }

  /** 
   * @override
   * @param { Student } student 
   */
  addStudent(student) {
    if (student.isAvailable(this._onlineCourse)) {
      this._onlineCourse.getStudents().push(student);
    } else {
      throw new Error("프라이빗 코스의 수강 권한이 없습니다.");
    }
  }

  /**
   * @override
   * @param { Student } student 
   * @param { string } review 
   */
  addReview(student, review) {
    if (this._onlineCourse.getStudents().includes(student)) {
      this._onlineCourse.getReviews().push(review);
    } else {
      throw new Error("프라이빗 코스를 수강하는 학생만 리뷰를 남길 수 있습니다.");
    }
  }
}

File: src/_03_BehavioralPattern/_03_08_State/java/resolveCircular/PrivateState.java
package _03_BehavioralPattern._03_08_State.java.resolveCircular;

public class PrivateState implements State {

  @Override
  public void addStudent(Student student, OnlineCourse onlineCourse) {
    if (student.isAvailable(onlineCourse)) {
      onlineCourse.addStudentDirectly(student);
    } else {
      throw new UnsupportedOperationException("[" + student.getName() + "] (은)는 수강 권한이 없습니다.");
    }
  }

  @Override
  public void addReview(Student student, String review, OnlineCourse onlineCourse) {
    if (onlineCourse.isRegisteredStudent(student)) {
      onlineCourse.addReviewDirectly(review);
    } else {
      throw new UnsupportedOperationException("수강중인 학생만 리뷰를 등록할 수 있습니다.");
    }
  }

  @Override
  public String toString() {
    return "비공개 상태";
  }
  
}
